Create activities table

crabsinger opened the pull request:
# Create an activities table to build an activities feed

- create database migrations
- setup hasura events
- create event_trigger and test

## Testing:
- create a contribution, add a user, create an epoch, and expect to see the activities table auto populated.
- also added tests for the createActivity file


## Questions:
- we wanted the activities table to hold `actor_profile_id` and `target_profile_id` but that data isn't always passed through, so we need to pass it through or fetch it later.
For users and epochs, we don't have the actor_profile....

## API:

Standard graphql using relations to expose nested activity data types

**Request**
```graphql
query Activities {
  activities {
    epoch {
      id
    }
    action
    user {
      address
      entrance
    }
    contribution {
      description
      id
    }
  }
}
```

**Response**

crabsinger commented on hasura/migrations/default/1675459146542_create_activities_table/up.sql:
```diff
@@ -0,0 +1,45 @@
+CREATE TABLE "public"."activities" (
```
@levity i dont need you to fully review this code (unless you want to) but since this is a big new DB architecture, I invite your feedback and thoughts before this is merged as its much easier to incorporate changes earlier rather than later.

levity replied:
this looks good to me -- i can see that avoiding more exotic polymorphism patterns would probably be most compatible with Hasura, and also GraphQL generally -- it seems like [the spec is still in flux](https://github.com/graphql/graphql-spec/issues?q=polymorphism+is%3Aopen)

CryptoGraffe commented on api-test/hasura/event_triggers/createActivity.payloads.ts:
```diff
@@ -0,0 +1,107 @@
+export const payloads = {
```
is there a better naming convention so that this is obviously for a test?

CryptoGraffe replied:
maybe even a subdirectory

CryptoGraffe commented on hasura/metadata/databases/default/tables/public_contributions.yaml:
```diff
@@ -90,6 +90,26 @@ select_permissions:
               _is_null: true
       allow_aggregations: true
 event_triggers:
+  - name: contributionInsertCreateActivity
```
might be nice to rename all these to have activity in the front so they are grouped in the UI

crabsinger replied:
ok, how about we name the trigger `activityHandler` since we probably will want to support more than just update in that same file.

and then we can rename these to `activityHandlerContributionInsert`

how's that sound?

CryptoGraffe commented on hasura/migrations/default/1675459146542_create_activities_table/up.sql:
```diff
@@ -0,0 +1,45 @@
+    "updated_at" timestamptz NOT NULL DEFAULT now(),
+
+  PRIMARY KEY ("id") ,
+  FOREIGN KEY ("circle_id") REFERENCES "public"."circles"("id") ON UPDATE CASCADE ON DELETE set null,
```
If we are going to cascade (vs restrict) perhaps we should just delete this activity instead of orphan it? idk

crabsinger replied:
yeah, that sounds reasonable. I've updated this to cascade deletes.

levity commented on api-lib/gql/mutations.ts:
```diff
@@ -533,3 +533,15 @@ export async function insertInteractionEvents(
   return;
 }
+
+export async function insertActivity(
```
same comment here as in `api-lib/gql/queries`... actually, in this case, the mutation is so straightforward i wouldn't bother splitting it out into a separate function at all

levity commented on api-lib/gql/queries.ts:
```diff
@@ -63,6 +63,50 @@ export async function getCircles() {
+        },
+      ],
+    },
+    {
```
please avoid unnecessary newlines where possible

levity commented on hasura/migrations/default/1675459146542_create_activities_table/up.sql:
```diff
@@ -0,0 +1,45 @@
+    "epoch_id" bigint,
+    "user_id" bigint,
+
+    "action" varchar(100) NOT NULL,
```
this field needs stricter validation -- it should fail to insert if it isn't in a specified set of strings

crabsinger replied:
talked with graffe - holding off on this for now.

crabsinger commented on api-lib/gql/queries.ts:
```diff
@@ -63,6 +63,50 @@ export async function getCircles() {
   ).circles;
 }
 
+export async function getOrgAndProfile(user_id: number, circle_id: number) {
```
done

crabsinger commented on api-test/hasura/event_triggers/createActivity.test.ts:
```diff
@@ -0,0 +1,149 @@
+  insertActivity: jest.fn(),
+}));
+
+describe('Create Activity Event Trigger', () => {
```
ok, yeah, this output is plenty nice:

<img width="744" alt="Screenshot 2023-02-15 at 9 31 04 AM" src="https://user-images.githubusercontent.com/83605543/219107389-78687154-8e70-43bd-8891-e6ad93b7e1fc.png">

crabsinger commented on api/hasura/event_triggers/eventManager.ts:
```diff
@@ -24,6 +25,9 @@ import { verifyHasuraRequestMiddleware } from '../../../api-lib/validate';
 type HandlerDict = { [handlerName: string]: VercelApiHandler };
 
 const HANDLERS: HandlerDict = {
+  usersInsertCreateActivity: createActivity,
```
1 handler, but three event_triggers that fire that same handler. each trigger is tied to a table action "contribution_insert" or something like that. We're going to eventually have a lot of triggers, but hopefully only one handler. @levity

levity replied:
what i meant was, is it possible to define a single handler here named `actvityHandler` and have all three new `event_triggers` YAML sections all refer to that same handler with 
```diff
-  - name: activityHandlerContributionInsert
+  - name: activityHandler
``` 

but i just checked the docs and the event trigger name has to be unique, i guess 🤷

crabsinger commented:
Ok, addressed all the feedback. I think this is ready to merge. I think our types are a bit weird right now, as I am seeing some diffs with `yarn generate` around GuildInfoInput added and Discord stuff being removed... so not sure about that...

levity commented on api-lib/event_triggers/activity/activityHandler.ts:
```diff
@@ -0,0 +1,142 @@
+import type { VercelRequest, VercelResponse } from '@vercel/node';
```
now that you've created a folder named `activity`, having the files also be prefixed with `activity` seems redundant to me, and none of the other files have `Handler` in their name, so how about renaming to `index.ts` and `mutations.ts`?

crabsinger replied:
ok, addressed all the feedback.
